### PR TITLE
feat(EMS-824): Insurance - Feedback - Feedback page and email

### PR DESCRIPTION
**Pull request by Zainzzkk**
# Changes made:
* Added feedback page for insurance
* Added e2e tests and unit tests
* Deactivated header feedback link temporarily

### CODE REVIEW

File: e2e-tests/constants/routes/insurance/index.js [hunk 1]
```diff
@@ -41,4 +41,6 @@ export const INSURANCE_ROUTES = {
   DECLARATIONS,
   CHECK_YOUR_ANSWERS,
   APPLICATION_SUBMITTED: '/application-submitted',
+  FEEDBACK: `${INSURANCE_ROOT}/give-feedback`,
+  FEEDBACK_CONFIRMATION: `${INSURANCE_ROOT}/feedback-sent`,
```
Comment (ttbarnes) [inline e2e-tests/constants/routes/insurance/index.js:45]: `FEEDBACK_SENT`?

File: e2e-tests/cypress/e2e/journeys/insurance/feedback/submit-feedback.spec.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      feedbackPage[SATISFACTION][SATISFIED].input().click();
+      cy.keyboardInput(feedbackPage[IMPROVEMENT].input(), 'test');
+      cy.keyboardInput(feedbackPage[OTHER_COMMENTS].input(), 'test');
+      submitButton().click();
+
+      cy.url().should('eq', feedbackConfirmationUrl);
+    });
+  });
+
+  describe('when clicking ', () => {
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/journeys/insurance/feedback/submit-feedback.spec.js:45]: `form submission`?

Could we also add a test for empty form submission?

```js
describe('when submitting an empty form', () => {
  it('should NOT render validation errorrs', () => {
  });
});
```

File: e2e-tests/cypress/e2e/journeys/insurance/feedback/validation/feedback-validation.spec.js [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  before(() => {
+    cy.navigateToUrl(startUrl);
+    partials.phaseBanner.feedbackLink().click();
+  });
+
+  beforeEach(() => {
+    cy.saveSession();
+  });
+
+  describe('when validation errors are present', () => {
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/journeys/insurance/feedback/validation/feedback-validation.spec.js:36]: I think we can remove this line doesn't seem needed

File: e2e-tests/cypress/e2e/pages/insurance/feedback/feedbackPage.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    hint: () => cy.get(`[data-cy="${IMPROVEMENT}-hint"]`),
+    input: () => cy.get(`[data-cy="${IMPROVEMENT}"]`),
+    errorMessage: () => cy.get(`[data-cy="${IMPROVEMENT}-error"]`),
+  },
+  [OTHER_COMMENTS]: {
+    label: () => cy.get(`[data-cy="${OTHER_COMMENTS}-label"]`),
+    hint: () => cy.get(`[data-cy="${OTHER_COMMENTS}-hint"]`),
+    input: () => cy.get(`[data-cy="${OTHER_COMMENTS}"]`),
+    errorMessage: () => cy.get(`[data-cy="${OTHER_COMMENTS}-error"]`),
+  },
```
Comment (ttbarnes) [inline e2e-tests/cypress/e2e/pages/insurance/feedback/feedbackPage.js:52]: How about one function for all of these fields? Started doing this in other places, DRY 😊 

```js
field: (fieldId) => ({ ...
```

File: src/api/custom-resolvers/index.ts [hunk 5]
```diff
@@ -18,4 +19,5 @@ export {
   addAndGetOTP,
   deleteApplicationByReferenceNumber,
   submitApplication,
+  sendEmailInsuranceFeedback,
```
Comment (ttbarnes) [inline src/api/custom-resolvers/index.ts:22]: Could this be `sendEmailFeedback` unless there are differences between quote and insurance?

File: src/api/custom-resolvers/send-email-insurance-feedback.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import { InsuranceFeedbackVariables, InsuranceFeedbackResponse } from '../types';
+
+/**
+ * sendEmailInsuranceFeedback
+ * - Sends email with feedback from insurance tool
+ * @param {Object} GraphQL root variables
+ * @param {Object} GraphQL variables for the insurance feedback
+ * @returns {Object} Object with success flag
+ */
+const sendEmailInsuranceFeedback = async (root: any, variables: InsuranceFeedbackVariables): Promise<InsuranceFeedbackResponse> => {
```
Comment (ttbarnes) [inline src/api/custom-resolvers/send-email-insurance-feedback.ts:11]: Unit test?

File: src/api/custom-schema.ts [hunk 7]
```diff
@@ -154,6 +155,10 @@ export const extendGraphqlSchema = (schema: GraphQLSchema) =>
         accountId: String
       }
 
+      type sendEmailInsuranceFeedbackResponse {
+        success: Boolean!
+      }
```
Comment (ttbarnes) [inline src/api/custom-schema.ts:160]: Can we call this `SuccessResponse`? In a PR that may or may not be merged before this - cleaned up some of these duplicated sucess instances

File: src/ui/server/routes/insurance/index.ts [hunk 8]
```diff
@@ -47,4 +49,10 @@ insuranceRouter.use(INSURANCE_ROOT, insuranceCheckYourAnswersRouter);
 
 insuranceRouter.get(`${INSURANCE_ROOT}/:referenceNumber${INSURANCE_ROUTES.APPLICATION_SUBMITTED}`, applicationSubmittedGet);
 
+insuranceRouter.get(INSURANCE_ROUTES.FEEDBACK, feedbackGet);
+insuranceRouter.post(INSURANCE_ROUTES.FEEDBACK, feedbackPost);
+
+insuranceRouter.get(INSURANCE_ROUTES.FEEDBACK_CONFIRMATION, feedbackConfirmationGet);
+insuranceRouter.post(INSURANCE_ROUTES.FEEDBACK_CONFIRMATION, feedbackConfirmationPost);
```
Comment (ttbarnes) [inline src/ui/server/routes/insurance/index.ts:56]: Do we need a POST for the sucess/confirmation page? If there is a button link perhaps it could just be a button link?

File: src/ui/templates/partials/phase-banner.njk [hunk 9]
```diff
@@ -8,6 +8,6 @@
     attributes: {
       "data-cy": "phase-banner"
     },
-    html: '<span data-cy="phase-banner-text">This is a new service - your <a class="govuk-link" href="https://forms.office.com/r/TacytrRCgJ" data-cy="phase-banner-feedback-link">feedback</a> will help us to improve it.</span>'
+    html: '<span data-cy="phase-banner-text">This is a new service - your <a class="govuk-link" href="/insurance/give-feedback" data-cy="phase-banner-feedback-link">feedback</a> will help us to improve it.</span>'
```
Comment (ttbarnes) [inline src/ui/templates/partials/phase-banner.njk:11]: Shouldn't the href be coming for page variables? `FEEDBACK_ROUTE`? Otherwise this change breaks the quote feedback

File: e2e-tests/cypress/e2e/journeys/insurance/feedback/validation/feedback-validation.spec.js (hunk 3 above)
Comment (Zainzzkk) [reply e2e-tests/cypress/e2e/journeys/insurance/feedback/validation/feedback-validation.spec.js:36]: I thought to use it but it will double the duration of the test because we are testing 2 errors at the same time

File: src/ui/templates/partials/phase-banner.njk (hunk 9 above)
Comment (Zainzzkk) [reply src/ui/templates/partials/phase-banner.njk:11]: this is fixed in the next PR

File: src/api/custom-resolvers/index.ts (hunk 5 above)
Comment (Zainzzkk) [reply src/api/custom-resolvers/index.ts:22]: Most likely will be different templates as different fields as more fields on the feedback form for quote

File: src/api/custom-resolvers/send-email-insurance-feedback.test.ts [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+  let sendInsuranceFeedbackEmailSpy = jest.fn();
+
+  afterAll(() => {
+    jest.resetAllMocks();
+  });
+
+  beforeEach(async () => {
+    variables = {
+      ...mockInsuranceFeedback,
+    };
```
Comment (ttbarnes) [inline src/api/custom-resolvers/send-email-insurance-feedback.test.ts:23]: Looks like this could be defined at the top one time
```
const variables = mockInsuranceFeedback as InsuranceFeedbackVariables;
```

File: src/ui/server/controllers/insurance/feedback/feedback-confirmation/index.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+ * @returns {Express.Response.render} renders insurance feedback page
+ */
+const get = (req: Request, res: Response) => {
+  try {
+    return res.render(TEMPLATE, {
+      ...insuranceCorePageVariables({
+        PAGE_CONTENT_STRINGS: FEEDBACK_SENT_PAGE,
+        BACK_LINK: req.headers.referer,
+      }),
+      BACK_TO_SERVICE_URL: req.flash('feedbackFrom'),
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/feedback/feedback-confirmation/index.ts:24]: `feedbackForm` confused me a bit - shall we call it maybe `feedbackOriginUrl`?

File: src/ui/server/controllers/insurance/feedback/feedback-form/index.test.ts [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+  describe('pageVariables', () => {
+    it('should have correct properties', () => {
+      const result = pageVariables();
+
+      const expected = {
+        FIELDS: {
+          SATISFACTION: {
+            ID: SATISFACTION,
+            VALUE_ID: { VERY_SATISFIED, SATISFIED, NEITHER, DISSATISFIED, VERY_DISSATISIFED },
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/feedback/feedback-form/index.test.ts:44]: What's the reason for `VALUE_ID`? Could we name it `OPTIONS` object instead so it's aligned with other radios?

File: src/ui/server/controllers/insurance/feedback/feedback-form/index.test.ts [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+        [IMPROVEMENT]: 'a'.repeat(MAXIMUM + 1),
+        [OTHER_COMMENTS]: '',
+      };
+
+      it('should redirect to next page', async () => {
+        req.body = body;
+
+        await post(req, res);
+
+        await post(req, res);
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/feedback/feedback-form/index.test.ts:159]: double post?

File: src/ui/server/controllers/insurance/feedback/feedback-form/index.ts [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        submittedValues: body,
+      });
+    }
+
+    if (objectHasKeysAndValues(feedback)) {
+      const emailVariables = {
+        // satisfaction will be null if not selected so set as empty string if null
+        [SATISFACTION]: feedback[SATISFACTION] ?? '',
+        [IMPROVEMENT]: feedback[IMPROVEMENT],
+        otherComments: feedback[OTHER_COMMENTS],
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/feedback/feedback-form/index.ts:95]: Variable for `otherComments`? Also do we need the `??` for `SATISFACTION`?

File: src/ui/server/controllers/insurance/feedback/feedback-form/validation/rules/improve-service.ts [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+  FEEDBACK: { IMPROVEMENT: FIELD_ID },
+} = FIELD_IDS;
+
+const { [FIELD_ID]: ERROR_MESSAGE } = ERROR_MESSAGES;
+
+export const MAXIMUM = 1200;
+
+/**
+ * validates improve service field
+ * checks if response has been provided
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/feedback/feedback-form/validation/rules/improve-service.ts:17]: "answer" instead of "response"?

File: src/ui/server/controllers/insurance/feedback/feedback-form/validation/rules/other-comments.ts [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+  FEEDBACK: { OTHER_COMMENTS: FIELD_ID },
+} = FIELD_IDS;
+
+const { [FIELD_ID]: ERROR_MESSAGE } = ERROR_MESSAGES;
+
+export const MAXIMUM = 1200;
+
+/**
+ * validates other comments field
+ * checks if response has been provided
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/feedback/feedback-form/validation/rules/other-comments.ts:17]: same as above

File: src/ui/server/helpers/object/index.test.ts [hunk 17]
```diff
@@ -25,6 +25,14 @@ describe('server/helpers/object', () => {
         expect(result).toEqual(false);
       });
     });
+
+    describe('when there is no object', () => {
```
Comment (ttbarnes) [inline src/ui/server/helpers/object/index.test.ts:29]: Test description says no object, but an object is provided 😝

File: src/ui/server/shared-validation/feedback/index.ts [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+ * @param {string} errorMessage
+ * @param {object} errors
+ * @param {number} maximum
+ * @returns {object} errors
+ */
+const validateFeedbackInput = (fieldBody: string, fieldId: string, errorMessage: string, errors: object, maximum: number) => {
+  let updatedErrors = errors;
+
+  // is over maximum number of characters then will be false
+  if (isAboveMaxLength(fieldBody, maximum)) {
```
Comment (ttbarnes) [inline src/ui/server/shared-validation/feedback/index.ts:17]: Nice, could this shared validation be called `max-length`? Then it can be used for any max length fields instead of being labelled specific for feedback

File: src/ui/templates/insurance/feedback-confirmation.njk [hunk 19]
```diff
@@ -0,0 +1,29 @@
+{% extends 'index.njk' %}
+{% from "govuk/components/button/macro.njk" import govukButton %}
+
+{% block pageTitle %}
+  {{ CONTENT_STRINGS.PAGE_TITLE }}
+{% endblock %}
+
+{% block content %}
+
+<h1 class="govuk-heading-l" data-cy="heading">{{ CONTENT_STRINGS.PAGE_TITLE }}</h1>
```
Comment (ttbarnes) [inline src/ui/templates/insurance/feedback-confirmation.njk:10]: indentation

File: src/ui/server/helpers/object/index.test.ts (hunk 17 above)
Comment (Zainzzkk) [reply src/ui/server/helpers/object/index.test.ts:29]: removed the test, since eslint kicks off

**Review by ttbarnes — APPROVED** [review]

**Review by ttbarnes — APPROVED** [review]
